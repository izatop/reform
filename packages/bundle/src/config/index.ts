export * from "./interfaces.js";
export * from "./JSONConfig.js";
