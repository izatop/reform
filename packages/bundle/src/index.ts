export * from "./build";
export * from "./package";
export * from "./plugins";
export * from "./internal";
