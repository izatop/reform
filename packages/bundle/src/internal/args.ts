import {assert} from ".";
import logger from "./logger";

export interface IArgumentList {
    print?: true;
    path: string;
    watch: boolean;
    serve?: boolean;
    mode: "production" | "development";
    ids: string[];

    readonly isProduction: boolean;
    readonly isDevelopment: boolean;
    readonly verbose: boolean;
    readonly silent: boolean;
}

export type InputArgumentList = {
    -readonly [K in keyof IArgumentList]: IArgumentList[K];
};

export function getArgumentList(root?: string): IArgumentList {
    const argumentList: InputArgumentList = {
        get isDevelopment() {
            return !this.isProduction;
        },
        get isProduction() {
            return this.mode !== "development";
        },
        ids: [],
        path: root ?? process.cwd(),
        mode: "production",
        watch: false,
        verbose: false,
        silent: false,
    };

    const {argv: [...args]} = process;
    while (args.length > 0) {
        const arg = args.shift();
        switch (arg) {
            case "-p":
            case "--print":
                argumentList.print = true;
                break;

            case "-v":
            case "--verbose":
                argumentList.verbose = true;
                logger.verbose = true;
                break;
            case "-s":
            case "--silent":
                argumentList.silent = true;
                logger.silent = true;
                break;
            case "--serve":
                argumentList.serve = true;
                argumentList.watch = true;
                break;

            case "--dev":
                argumentList.mode = "development";
                break;

            case "-w":
            case "--watch":
                argumentList.watch = true;
                break;
            case "-i":
            case "--id":
                assert(args[0], "The bundle ID expected after --id/-i option");
                argumentList.ids.push(args[0]);
                args.shift();
        }
    }

    return argumentList;
}
