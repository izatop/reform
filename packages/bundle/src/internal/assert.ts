export class AssertionError extends Error {
    constructor(message: string, details?: unknown) {
        super(message.concat("\n\n", JSON.stringify(details, null, 2)));
    }
}

export function assert(expr: unknown, message = "Assertion error", details?: unknown): asserts expr {
    if (!expr) {
        throw new AssertionError(message, details);
    }
}
