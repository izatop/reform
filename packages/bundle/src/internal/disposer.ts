import {trySafe} from "../internal/index.js";
import logger from "../internal/logger.js";

export type Disposable = () => unknown;

const disposables: Disposable[] = [];

export class Disposer {
    public static add(disposable: Disposable) {
        disposables.push(disposable);
    }

    public static dispose() {
        logger.debug("Dispose", "listeners -> %d", disposables.length);
        disposables.forEach((fn) => trySafe(this, fn));
    }
}
