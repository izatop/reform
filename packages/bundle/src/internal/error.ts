export function isError(error: unknown): error is Error {
    return error instanceof Error;
}

export function withError<T>(error: unknown, fn: (error: Error) => T): T | undefined;
export function withError<T>(error: unknown, fn: (error: Error) => T, or: () => T): T;
export function withError<T>(error: unknown, fn: (error: Error) => T, or?: () => T): T | undefined {
    if (isError(error)) {
        return fn(error);
    }

    return or?.();
}
