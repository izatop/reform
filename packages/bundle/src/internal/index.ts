export * from "./args";
export * from "./assert";
export * from "./module";
export * from "./promise";
export * from "./path";
export * from "./main";
export * from "./logger";
