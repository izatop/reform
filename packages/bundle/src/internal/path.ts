import {ok} from "assert";
import * as path from "path";

export function resolveAt(base: string, resource: string): string {
    return path.join(base, resource);
}

export function resolveStrictAt(base: string, resource: unknown, message: string) {
    ok(typeof resource === "string", message);
    return resolveAt(base, resource);
}

export function relativeTo(from: string, to: string): string {
    return path.relative(from, to);
}
