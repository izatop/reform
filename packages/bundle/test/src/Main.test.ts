import {Directory, exit, isPluginCtor, load} from "../../src";
import {TestPlugin} from "./plugin/TestPlugin";
import {resolve} from "path";
import {BuildContext} from "../../src/build/BuildContext";
import {getArgumentList} from "../../src/internal";

test("Main Test", async () => {
    const id = resolve(__dirname, "plugin/TestPlugin");
    const ctx = new BuildContext({
        id,
        args: getArgumentList(),
        base: new Directory(__dirname, "/src"),
        build: new Directory(__dirname, "/build"),
        entries: ["index.html"],
        format: "esm",
        platform: "browser",
    });

    expect(isPluginCtor(TestPlugin)).toBe(true);
    expect(load(id, ctx, {})).toMatchSnapshot();
    exit(0);
});
