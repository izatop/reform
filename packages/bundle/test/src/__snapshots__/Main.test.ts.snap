// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Test 1`] = `
Object {
  "name": "TestPlugin",
  "setup": [Function],
}
`;
